packages/dds/matrix/src/matrix.ts

```diff
@@ -357,16 +357,22 @@ export class SharedMatrix<T extends Serializable = Serializable>
     /** @internal */ public _undoRemoveRows(segment: ISegment) {
         const original = segment as PermutationSegment;
 
-        // (Re)insert the removed number of columns at the original position.
+        // (Re)insert the removed number of rows at the original position.
         const rowStart = this.rows.getPosition(original);
         this.insertRows(rowStart, original.cachedLength);
 
-        // Transfer handles from the original segment to the newly inserted segment.
-        // (This allows us to use getCell(..) below to read the previous cell values)
+        // Transfer handles from the original segment to the newly inserted empty segment.
         const inserted = this.rows.getContainingSegment(rowStart).segment as PermutationSegment;
         original.transferHandlesTo(inserted);
 
-        // Generate setCell ops for each populated cell in the reinserted cols.
+        // Invalidate the handleCache in case it was populated during the 'rowsChanged'
+        // callback, which occurs before the handle span is populated.
+        this.rows.handleCache.itemsChanged(
+            rowStart,
+            /* removedCount: */ 0,
+            /* insertedCount: */ inserted.cachedLength);
+
```

Comment:
(This is the bug fix.)

Reply:
In the permutation vector `onDelta` the handleCache gets invalided, so was this a case where the code path doesn't end up calling `onDelta`, or that something happens before `onDelta` is called?

Reply:
It was a fun edge case that went like this:

* onDelta is called
    * onDelta clears the cache
    * onDelta invokes row/colChanged
        * Inside row/colChanged, the app reads cells from the inserted row/col (populating the cache with empty)
* Handles are transferred to the new segment
* cellChanged is called
    * App reads same cells, which the handle cache says is still an empty segment.

It's really tempting to transfer the handles prior to inserting the new segment so the inserted row/cols are initially populated in the row/colChanged callbacks, but that only works on the local client.